File: app/src/main/java/com/ramez/shopp/Fragments/RegisterFragment.java
package com.ramez.shopp.Fragments;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.text.method.HideReturnsTransformationMethod;
import android.text.method.PasswordTransformationMethod;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.viewpager.widget.ViewPager;

import com.github.dhaval2404.form_validation.rule.EqualRule;
import com.github.dhaval2404.form_validation.rule.NonEmptyRule;
import com.github.dhaval2404.form_validation.validation.FormValidator;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.FirebaseInstanceIdReceiver;
import com.onesignal.OneSignal;
import com.ramez.shopp.Activities.ChooseCityActivity;
import com.ramez.shopp.Activities.ConditionActivity;
import com.ramez.shopp.Activities.ConfirmActivity;
import com.ramez.shopp.Activities.TermsActivity;
import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.GlobalData;
import com.ramez.shopp.Classes.OtpModel;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Models.LocalModel;
import com.ramez.shopp.Models.LoginResultModel;
import com.ramez.shopp.Models.MemberModel;
import com.ramez.shopp.R;
import com.ramez.shopp.Utils.NumberHandler;
import com.ramez.shopp.Utils.SharedPManger;
import com.ramez.shopp.databinding.FragmentRegisterBinding;

public class RegisterFragment extends FragmentBase {
    String FCMToken;
    String CountryCode = "+966";
    boolean select_country = false;
    String country_name = "BH";
    String city_id = "7263";
    String prefix = "973";
    SharedPManger sharedPManger;
    private FragmentRegisterBinding binding;
    private ViewPager viewPager;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentRegisterBinding.inflate(inflater, container, false);
        View view = binding.getRoot();
        viewPager = container.findViewById(R.id.viewPager);

        getDeviceToken();
        sharedPManger = new SharedPManger(getActivityy());

        binding.loginBut.setOnClickListener(view1 -> {
            startLogin();


        });


        binding.showPassBut.setOnClickListener(view1 -> {

            if (binding.edtPassword.getTransformationMethod().equals(PasswordTransformationMethod.getInstance())) {
                ((ImageView) (view1)).setImageResource(R.drawable.ic_visibility_off);
                binding.edtPassword.setTransformationMethod(HideReturnsTransformationMethod.getInstance());
            } else {
                ((ImageView) (view1)).setImageResource(R.drawable.ic_visibility);
                binding.edtPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
            }
        });



        binding.showConfirmPassBut.setOnClickListener(view1 -> {

            if (binding.edtConfirmPassword.getTransformationMethod().equals(PasswordTransformationMethod.getInstance())) {
                ((ImageView) (view1)).setImageResource(R.drawable.ic_visibility_off);
                binding.edtConfirmPassword.setTransformationMethod(HideReturnsTransformationMethod.getInstance());
            } else {
                ((ImageView) (view1)).setImageResource(R.drawable.ic_visibility);
                binding.edtConfirmPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
            }
        });

        binding.registerBut.setOnClickListener(view1 -> {
            if (isValidForm()) {
                RegisterUser();

            }
        });

        binding.privacyBut.setOnClickListener(view1 -> {
            startConditionActivity();


        });

        binding.termsBut.setOnClickListener(view1 -> {
            startTermsActivity();


        });

        return view;
    }

    private void RegisterUser() {
        final String mobileStr = NumberHandler.arabicToDecimal(binding.edtPhoneNumber.getText().toString());
        final String passwordStr = NumberHandler.arabicToDecimal(binding.edtPassword.getText().toString());
        final String nameStr = NumberHandler.arabicToDecimal(binding.edtFirstName.getText().toString());
        final String emailStr = NumberHandler.arabicToDecimal(binding.edtEmail.getText().toString());
        LocalModel localModel = UtilityApp.getLocalData();

        country_name = localModel.getShortname();
        CountryCode = String.valueOf(localModel.getPhonecode());
        city_id = localModel.getCityId();

        MemberModel memberModel = new MemberModel();
        memberModel.setMobileNumber(mobileStr);
        memberModel.setPassword(passwordStr);
        memberModel.setName(nameStr);
        memberModel.setEmail(emailStr);
        memberModel.setCity(city_id);
        memberModel.setCountry(country_name);
        memberModel.setDeviceToken(FCMToken);
        memberModel.setDeviceId(UtilityApp.getUnique());
        memberModel.setDeviceType(Constants.deviceType);
        memberModel.setPrefix(CountryCode);
        memberModel.setUserType(Constants.user_type);

        GlobalData.progressDialog(getActivityy(), R.string.register, R.string.please_wait_register);

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            GlobalData.hideProgressDialog();
            LoginResultModel result = (LoginResultModel) obj;
            if (func.equals(Constants.ERROR)) {
                String message = getString(R.string.fail_register);
                if (result != null && result.getMessage() != null) {
                    message = result.getMessage();
                }
                GlobalData.errorDialog(getActivityy(), R.string.register, message);
            } else if (func.equals(Constants.NO_CONNECTION)) {
                GlobalData.Toast(getActivityy(), R.string.no_internet_connection);
            } else {
                if (IsSuccess) {
                        Log.i("TAG", "Log getStatus " + result.getStatus());

                        if (result.getStatus() == 200) {
                            MemberModel user = result.data;
                            if(user!=null){
                                user.setUserType(Constants.user_type);

                            }

                            UtilityApp.setUserData(user);
                            SendOtp(mobileStr, passwordStr);

                        } else {
                            String message = getString(R.string.fail_register);
                            if (result.getMessage() != null) {
                                message = result.getMessage();
                            }
                            GlobalData.errorDialog(getActivityy(), R.string.register, message);


                        }


                } else {
                    Toast(getString(R.string.fail_register));

                }
            }


        }).RegisterHandle(memberModel);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void startLogin() {
        viewPager.setCurrentItem(1);

    }

    private final boolean isValidForm() {

        FormValidator formValidator = FormValidator.Companion.getInstance();

        return formValidator.addField(binding.edtFirstName, new NonEmptyRule(getString(R.string.enter_name))).addField(binding.edtPhoneNumber, new NonEmptyRule(getString(R.string.enter_phone_number))).
                addField(binding.edtPassword, new NonEmptyRule(getString(R.string.enter_password))).

                addField(binding.edtConfirmPassword, new NonEmptyRule(getString(R.string.enter_confirm_password)), new EqualRule(String.valueOf(binding.edtPassword.getText()), R.string.password_confirm_not_match)).validate();

    }

    private void getDeviceToken() {

        FCMToken = UtilityApp.getFCMToken();
        if (FCMToken == null) {

            OneSignal.idsAvailable((userId, registrationId) -> {
                Log.d("debug", "Log User:" + userId);
                if (registrationId != null)
                    FCMToken=OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId();
                UtilityApp.setFCMToken(FCMToken);

                Log.d("debug", "Log token one signal first :" + OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId());
                Log.d("debug", "Log token firebase:" + UtilityApp.getFCMToken());

            });

        }

    }


    private void GoToChooseCity() {
        Intent intent = new Intent(getActivityy(), ChooseCityActivity.class);
        startActivity(intent);
        getActivityy().finish();

    }


    public void SendOtp(String mobile, String password) {
        final String mobileStr = NumberHandler.arabicToDecimal(binding.edtPhoneNumber.getText().toString());
        new DataFeacher(false, (obj, func, IsSuccess) -> {
            if (func.equals(Constants.ERROR)) {
                Toast(R.string.error_in_data);
            } else if (func.equals(Constants.FAIL)) {
                Toast(R.string.fail_to_sen_otp);
            } else {
                if (IsSuccess) {
                    OtpModel otpModel = (OtpModel) obj;
                    if (otpModel.getData() != null) {
                        Log.i("TAG", "Log otp " + otpModel.getData());

                    }
                    Intent intent = new Intent(getActivityy(), ConfirmActivity.class);
                    intent.putExtra(Constants.KEY_MOBILE, mobileStr);
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);


                } else {
                    Toast(R.string.fail_to_sen_otp);
                }
            }

        }).sendOpt(mobile);
    }


    private void startTermsActivity() {
        Intent intent = new Intent(getActivityy(), TermsActivity.class);
        startActivity(intent);
    }

    private void startConditionActivity() {
        Intent intent = new Intent(getActivityy(), ConditionActivity.class);
        startActivity(intent);
    }

}

File: app/src/main/java/com/amour/shop/fragments/RegisterFragment.kt
package com.amour.shop.fragments

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.text.method.HideReturnsTransformationMethod
import android.text.method.PasswordTransformationMethod
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.viewpager.widget.ViewPager
import com.github.dhaval2404.form_validation.rule.EqualRule
import com.github.dhaval2404.form_validation.rule.NonEmptyRule
import com.github.dhaval2404.form_validation.validation.FormValidator
import com.onesignal.OneSignal
import com.amour.shop.classes.Constants
import com.amour.shop.classes.GlobalData
import com.amour.shop.activities.ConditionActivity
import com.amour.shop.activities.ConfirmActivity
import com.amour.shop.activities.TermsActivity
import com.amour.shop.ApiHandler.DataFeacher
import com.amour.shop.ApiHandler.DataFetcherCallBack
import com.amour.shop.Models.OtpModel
import com.amour.shop.classes.UtilityApp
import com.amour.shop.Models.LocalModel
import com.amour.shop.Models.MemberModel
import com.amour.shop.Models.RegisterResultModel
import com.amour.shop.R
import com.amour.shop.Utils.NumberHandler
import com.amour.shop.Utils.SharedPManger
import com.amour.shop.classes.Constants.MAIN_ACTIVITY_CLASS
import com.amour.shop.classes.GlobalData.errorDialog
import com.amour.shop.classes.GlobalData.getIntro
import com.amour.shop.classes.GlobalData.hideProgressDialog
import com.amour.shop.classes.GlobalData.progressDialog
import com.amour.shop.databinding.FragmentRegisterBinding
import com.amour.shop.BuildConfig


class RegisterFragment : FragmentBase() {
    var FCMToken: String? = null
    var countryCode = "+966"
    var countryName = "BH"
    var cityId = "7263"
    var sharedPManger: SharedPManger? = null
    private lateinit var binding: FragmentRegisterBinding
    private var viewPager: ViewPager? = null
    var localModel: LocalModel? = null

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentRegisterBinding.inflate(inflater, container, false)
        val view: View = binding.root
        viewPager = container!!.findViewById(R.id.viewPager)
        deviceToken
        sharedPManger = SharedPManger(activityy)
        localModel = UtilityApp.getLocalData()
        countryName = localModel?.shortname.toString()
        countryCode = localModel?.phonecode.toString()
        cityId = localModel?.cityId.toString()
        val intro = getIntro(countryCode)
        Log.i(javaClass.simpleName, "Log get  Intro $intro")
        binding.edtPhoneNumber.hint = intro
        initListeners()
        return view
    }

    private fun initListeners() {
        binding.loginBut.setOnClickListener { startLogin() }
        binding.showPassBut.setOnClickListener { view1 ->
            if (binding.edtPassword.transformationMethod
                        .equals(PasswordTransformationMethod.getInstance())
            ) {
                (view1 as ImageView).setImageResource(R.drawable.ic_visibility_off)
                binding.edtPassword.transformationMethod = HideReturnsTransformationMethod.getInstance()
            } else {
                (view1 as ImageView).setImageResource(R.drawable.ic_visibility)
                binding.edtPassword.transformationMethod = PasswordTransformationMethod.getInstance()
            }
        }
        binding.showConfirmPassBut.setOnClickListener { view1 ->
            if (binding.edtConfirmPassword.transformationMethod
                        .equals(PasswordTransformationMethod.getInstance())
            ) {
                (view1 as ImageView).setImageResource(R.drawable.ic_visibility_off)
                binding.edtConfirmPassword.transformationMethod =
                    HideReturnsTransformationMethod.getInstance()
            } else {
                (view1 as ImageView).setImageResource(R.drawable.ic_visibility)
                binding.edtConfirmPassword.transformationMethod = PasswordTransformationMethod.getInstance()
            }
        }
        binding.registerBut.setOnClickListener {
            if (isValidForm) {
                registerUser()
            }
        }
        binding.privacyBut.setOnClickListener { startConditionActivity() }
        binding.termsBut.setOnClickListener { startTermsActivity() }
        binding.skipButton.setOnClickListener { startMain() }
    }

    private fun registerUser() {
        val mobileStr = NumberHandler.arabicToDecimal(binding.edtPhoneNumber.text.toString())
        val passwordStr = NumberHandler.arabicToDecimal(binding.edtPassword.text.toString())
        val nameStr = NumberHandler.arabicToDecimal(binding.edtFirstName.text.toString())
        val memberModel = MemberModel()
        memberModel.mobileNumber = mobileStr
        memberModel.password = passwordStr
        memberModel.name = nameStr
        memberModel.email = ""
        memberModel.city = cityId.toInt()
        memberModel.country = countryName
        memberModel.deviceToken = FCMToken
        memberModel.deviceId = UtilityApp.unique
        memberModel.deviceType = Constants.deviceType
        memberModel.prefix = countryCode
        memberModel.userType = Constants.user_type
        progressDialog(activityy, R.string.register, R.string.please_wait_register)
        DataFeacher(false, object :
                DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                if (isVisible) {
                    hideProgressDialog()
                    val result = obj as RegisterResultModel?
                    if (func == Constants.ERROR) {
                        var message: String? = getString(R.string.fail_register)
                        if (result != null && result.message != null) {
                            message = result.message
                        }
                        errorDialog(activityy, R.string.register, message)
                    } else if (func == Constants.NO_CONNECTION) {
                        GlobalData.Toast(activityy, R.string.no_internet_connection)
                    } else {
                        if (IsSuccess) {
                            Log.i("TAG", "Log getStatus " + result!!.status)
                            if (result.status == 200) {
                                val user = result.data
                                if (user != null) {
                                    user.userType = Constants.user_type
                                }
                                UtilityApp.setUserData(user)

                                startLogin()

                            } else {
                                var message: String? = getString(R.string.fail_register)
                                if (result.message != null) {
                                    message = result.message
                                }
                                errorDialog(activityy, R.string.register, message)
                            }
                        } else {
                            Toast(getString(R.string.fail_register))
                        }
                    }
                }
            }
        }).registerHandle(memberModel)
    }

    private fun startLogin() {
        viewPager!!.currentItem = 1
    }

    private val isValidForm: Boolean
        get() {
            val formValidator = FormValidator.getInstance()
            return formValidator.addField(binding.edtFirstName, NonEmptyRule(getString(R.string.enter_name)))
                .addField(binding.edtPhoneNumber, NonEmptyRule(getString(R.string.enter_phone_number)))
                .addField(binding.edtPassword, NonEmptyRule(getString(R.string.enter_password))).addField(
                    binding.edtConfirmPassword,
                    NonEmptyRule(getString(R.string.enter_confirm_password)),
                    EqualRule(
                        java.lang.String.valueOf(binding.edtPassword.text),
                        R.string.password_confirm_not_match
                    )
                ).validate()
        }

    //            OneSignal.idsAvailable((userId, registrationId) -> {
    //                if (registrationId != null)
//                    FCMToken=OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId();
//                UtilityApp.setFCMToken(FCMToken);
//
//                Log.d("debug", "Log token one signal first :" + OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId());
//                Log.d("debug", "Log token firebase:" + UtilityApp.getFCMToken());
//
//            });
    private val deviceToken: Unit
        get() {
            FCMToken = UtilityApp.getFCMToken()
            if (FCMToken == null) {
                val FCMToken = OneSignal.getDeviceState()!!.userId
                UtilityApp.setFCMToken(FCMToken)

                //            OneSignal.idsAvailable((userId, registrationId) -> {
                Log.d("debug", "Log User:$FCMToken")
                //                if (registrationId != null)
                //                    FCMToken=OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId();
                //                UtilityApp.setFCMToken(FCMToken);
                //
                //                Log.d("debug", "Log token one signal first :" + OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId());
                //                Log.d("debug", "Log token firebase:" + UtilityApp.getFCMToken());
                //
                //            });
            }
        }


    fun SendOtp(mobile: String?, password: String?) {
        val mobileStr = NumberHandler.arabicToDecimal(binding.edtPhoneNumber.getText().toString())
        DataFeacher(false, object :
                DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                if (isVisible) {
                    if (func == Constants.ERROR) {
                        Toast(R.string.error_in_data)
                    } else if (func == Constants.FAIL) {
                        Toast(R.string.fail_to_sen_otp)
                    } else {
                        if (IsSuccess) {
                            val otpModel = obj as OtpModel
                            if (otpModel.status == 200) {
                                Log.i("TAG", "Log otp " + otpModel.data)
                                val intent = Intent(activityy, ConfirmActivity::class.java)
                                intent.putExtra(Constants.KEY_MOBILE, mobileStr)
                                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                                startActivity(intent)
                            }
                        } else {
                            Toast(R.string.fail_to_sen_otp)
                        }
                    }
                }
            }

        }).sendOpt(mobile ?: "")
    }

    private fun startTermsActivity() {
        val intent = Intent(activityy, TermsActivity::class.java)
        startActivity(intent)
    }

    private fun startConditionActivity() {
        val intent = Intent(activityy, ConditionActivity::class.java)
        startActivity(intent)
    }

    private fun startMain() {
        val intent = Intent(activityy, MAIN_ACTIVITY_CLASS)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(intent)
    }
}